Content.Server/Tools/Systems/WeldableSystem.cs

```diff
@@ -0,0 +1,152 @@
+using Content.Server.Tools.Components;
+using Content.Shared.Examine;
+using Content.Shared.Interaction;
+using Content.Shared.Tools.Components;
+
+namespace Content.Server.Tools.Systems;
+
+public sealed class WeldableSystem : EntitySystem
+{
+    [Dependency] private readonly ToolSystem _toolSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<WeldableComponent, InteractUsingEvent>(OnInteractUsing);
+        SubscribeLocalEvent<WeldableComponent, WeldFinishedEvent>(OnWeldFinished);
+        SubscribeLocalEvent<WeldableComponent, WeldCancelledEvent>(OnWeldCanceled);
+        SubscribeLocalEvent<WeldableComponent, ExaminedEvent>(OnExamine);
+    }
+
+    private void OnExamine(EntityUid uid, WeldableComponent component, ExaminedEvent args)
+    {
+        if (component.IsWelded && component.WeldedExamineMessage != null)
+            args.PushText(Loc.GetString(component.WeldedExamineMessage));
+    }
+
+    private void OnInteractUsing(EntityUid uid, WeldableComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.Handled = TryWeld(uid, args.Used, args.User, component);
+    }
+
+    private bool CanWeld(EntityUid uid, EntityUid tool, EntityUid user, WeldableComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        // Basic checks
+        if (!component.Weldable || component.BeingWelded)
+            return false;
+        if (!_toolSystem.HasQuality(tool, component.WeldingQuality))
+            return false;
+
+        // Other component systems
+        var attempt = new WeldableAttemptEvent(user, tool);
+        RaiseLocalEvent(uid, attempt);
+        if (attempt.Cancelled)
+            return false;
+
+        return true;
+    }
+
+    private bool TryWeld(EntityUid uid, EntityUid tool, EntityUid user, WeldableComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (!CanWeld(uid, tool, user, component))
+            return false;
+
+        component.BeingWelded = true;
+        _toolSystem.UseTool(tool, user, uid, component.FuelConsumption,
+            component.WeldingTime.Seconds, component.WeldingQuality,
+            new WeldFinishedEvent(user, tool), new WeldCancelledEvent(), uid);
```

Comment:
Would be good to have some way for other stuff to specify a modifier if it wants it to take longer.

Reply:
I could probably add a public function in `WeldableSystem` to change welding duration. Adding new event like `WeldingTimeModifierEvent` seems to be overkill for now.